In [1]:
# import dependencies
# data cleaning and manipulation 
import pandas as pd
import numpy as np

# machine learning
from sklearn.preprocessing import StandardScaler
import sklearn.linear_model as skl_lm
from sklearn import preprocessing
from sklearn import neighbors
from sklearn.metrics import confusion_matrix, classification_report, precision_score
from sklearn.model_selection import train_test_split

In [54]:
df = pd.read_csv('data/data.csv')
cols = ["id",
        "Unnamed: 32"]
df = df.drop(cols, axis=1)
print('Shape of the dataframe: ', df.shape)
# verify remaining columns
df.columns

Shape of the dataframe:  (569, 31)


Index(['diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [55]:
y = df.diagnosis.values

b, c = np.unique(y, return_inverse=True)
print(b)
print(c)

x_data = df.drop(['diagnosis'], axis = 1)

#Split the train and test dataset
x_train, x_test, y_train, y_test = train_test_split(x_data,c,test_size = 0.3 ,random_state=40)

['B' 'M']
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 0 1 1 1 1 1 1 1 1 0 1 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 1
 0 1 0 1 1 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0
 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1 1 0 1 1 0 0 1 0 0 1 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 1 1 1 0 1
 0 1 0 0 0 1 0 0 1 1 0 1 1 1 1 0 1 1 1 0 1 0 1 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0
 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 1 0 0
 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1
 1 0 1 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 1 0 0 0 0 0 1 0 0
 1 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 1 

In [56]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [57]:
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras import optimizers
from keras.wrappers.scikit_learn import KerasClassifier
from keras.callbacks import EarlyStopping
from keras.callbacks import ModelCheckpoint, TensorBoard

In [58]:
# Initialising the ANN
classifier = Sequential()
# Adding the input layer and the first hidden layer
classifier.add(Dense(output_dim=16, init='uniform', activation='relu', input_dim=30))
# Adding dropout to prevent overfitting
classifier.add(Dropout(p=0.1))
# Adding the second hidden layer
classifier.add(Dense(output_dim=16, init='uniform', activation='relu'))
# Adding dropout to prevent overfitting
classifier.add(Dropout(p=0.1))
# Adding the third hidden layer
classifier.add(Dense(output_dim=16, init='uniform', activation='relu'))
# Adding dropout to prevent overfitting
classifier.add(Dropout(p=0.1))
# Adding the output layer
classifier.add(Dense(output_dim=1, init='uniform', activation='sigmoid'))

# Compiling the ANN
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=30, units=16, kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:6: UserWarning: Update your `Dropout` call to the Keras 2 API: `Dropout(rate=0.1)`
  
C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:8: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=16, kernel_initializer="uniform")`
  
C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:10: UserWarning: Update your `Dropout` call to the Keras 2 API: `Dropout(rate=0.1)`
  # Remove the CWD from sys.path while we load stuff.
C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:12:

In [59]:
filepath = "weights.hdf5"
checkpoint = ModelCheckpoint(filepath, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
tboard = TensorBoard(log_dir='./logs', write_graph=True, write_images=True)
callbacks_list = [checkpoint, tboard]

classifier.fit(x_train, y_train, batch_size=100, nb_epoch=150, validation_data=(x_test, y_test), callbacks=callbacks_list)

C:\Users\HEMANTH KUMAR\.conda\envs\miniproject-sem6\lib\site-packages\ipykernel_launcher.py:6: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Train on 398 samples, validate on 171 samples
Epoch 1/150
398/398 [==============================] - 1s 2ms/step - loss: 0.6930 - acc: 0.5503 - val_loss: 0.6924 - val_acc: 0.6725

Epoch 00001: val_acc improved from -inf to 0.67251, saving model to weights.hdf5
Epoch 2/150
398/398 [==============================] - 0s 75us/step - loss: 0.6925 - acc: 0.6080 - val_loss: 0.6916 - val_acc: 0.6725

Epoch 00002: val_acc did not improve from 0.67251
Epoch 3/150
398/398 [==============================] - 0s 28us/step - loss: 0.6918 - acc: 0.6080 - val_loss: 0.6906 - val_acc: 0.6725

Epoch 00003: val_acc did not improve from 0.67251
Epoch 4/150
398/398 [==============================] - 0s 30us/step - loss: 0.6911 - acc: 0.6080 - val_loss: 0.6893 - val_acc: 0.6725

Epoch 00004: val_acc did not improve from 0.67251
Epoch 5/150
398/398 [==============================] - 0s 63us/step - loss: 0.6901 - acc: 0.6080 - val_loss: 0.6877 - val_acc: 0.6725

Epoch 00005: val_acc did not improve from 0.67251

398/398 [==============================] - 0s 25us/step - loss: 0.0823 - acc: 0.9799 - val_loss: 0.0779 - val_acc: 0.9708

Epoch 00043: val_acc did not improve from 0.98830
Epoch 44/150
398/398 [==============================] - 0s 25us/step - loss: 0.0863 - acc: 0.9799 - val_loss: 0.0776 - val_acc: 0.9708

Epoch 00044: val_acc did not improve from 0.98830
Epoch 45/150
398/398 [==============================] - 0s 33us/step - loss: 0.0777 - acc: 0.9824 - val_loss: 0.0771 - val_acc: 0.9708

Epoch 00045: val_acc did not improve from 0.98830
Epoch 46/150
398/398 [==============================] - 0s 35us/step - loss: 0.0810 - acc: 0.9799 - val_loss: 0.0767 - val_acc: 0.9708

Epoch 00046: val_acc did not improve from 0.98830
Epoch 47/150
398/398 [==============================] - 0s 53us/step - loss: 0.0769 - acc: 0.9824 - val_loss: 0.0764 - val_acc: 0.9708

Epoch 00047: val_acc did not improve from 0.98830
Epoch 48/150
398/398 [==============================] - 0s 30us/step - loss: 0.0769

398/398 [==============================] - 0s 65us/step - loss: 0.0450 - acc: 0.9899 - val_loss: 0.0871 - val_acc: 0.9766

Epoch 00087: val_acc did not improve from 0.98830
Epoch 88/150
398/398 [==============================] - 0s 40us/step - loss: 0.0490 - acc: 0.9824 - val_loss: 0.0873 - val_acc: 0.9766

Epoch 00088: val_acc did not improve from 0.98830
Epoch 89/150
398/398 [==============================] - 0s 35us/step - loss: 0.0457 - acc: 0.9874 - val_loss: 0.0876 - val_acc: 0.9766

Epoch 00089: val_acc did not improve from 0.98830
Epoch 90/150
398/398 [==============================] - 0s 30us/step - loss: 0.0482 - acc: 0.9849 - val_loss: 0.0882 - val_acc: 0.9766

Epoch 00090: val_acc did not improve from 0.98830
Epoch 91/150
398/398 [==============================] - 0s 98us/step - loss: 0.0501 - acc: 0.9849 - val_loss: 0.0890 - val_acc: 0.9766

Epoch 00091: val_acc did not improve from 0.98830
Epoch 92/150
398/398 [==============================] - 0s 50us/step - loss: 0.0446

Epoch 131/150
398/398 [==============================] - 0s 25us/step - loss: 0.0320 - acc: 0.9899 - val_loss: 0.1113 - val_acc: 0.9766

Epoch 00131: val_acc did not improve from 0.98830
Epoch 132/150
398/398 [==============================] - 0s 25us/step - loss: 0.0323 - acc: 0.9899 - val_loss: 0.1125 - val_acc: 0.9766

Epoch 00132: val_acc did not improve from 0.98830
Epoch 133/150
398/398 [==============================] - 0s 28us/step - loss: 0.0347 - acc: 0.9874 - val_loss: 0.1127 - val_acc: 0.9766

Epoch 00133: val_acc did not improve from 0.98830
Epoch 134/150
398/398 [==============================] - 0s 23us/step - loss: 0.0306 - acc: 0.9899 - val_loss: 0.1124 - val_acc: 0.9766

Epoch 00134: val_acc did not improve from 0.98830
Epoch 135/150
398/398 [==============================] - 0s 28us/step - loss: 0.0310 - acc: 0.9899 - val_loss: 0.1126 - val_acc: 0.9766

Epoch 00135: val_acc did not improve from 0.98830
Epoch 136/150
398/398 [==============================] - 0s 33us/

### Our neural network model obtained an accuracy of 98.83% on the validation set.